gms/gossiper.cc

```diff
@@ -1298,7 +1299,7 @@ bool gossiper::is_gossip_only_member(inet_address endpoint) {
     if (!es) {
         return false;
     }
-    return !is_dead_state(*es) && !get_token_metadata_ptr()->is_normal_token_owner(endpoint);
+    return !is_dead_state(*es) && !get_token_metadata_ptr()->is_normal_token_owner(endpoint) && !is_seed(endpoint);
```

Comment:
Why do we add the !is_seed here? A node being a seed can be a gossip only member.

Reply:
Because the idea is that it stops being a gossip only member. What is a gossip only member anyway? It's a helper function used to decide if we can completely drop an ip address from the ping list. This comment for the patch explains why it is a bad idea.

Reply:
A gossip only member is a node who is not part of the ring but a node in the gossip. It is used to drop a node from gossip, e.g., a node which tries to bootstrap into the cluster but failed in the middle. The node is not part of the ring, gossip will drop it automatically after it stops update for a while.

It is not used to decide whom gossip will talk with.

Reply:
This explanation is contradictory. First of all, with IP address changes gossip can not identify which node is part of the ring and which isn't. The ring is associated with host ids, not ip addresses, and gossip's view of the node is based on ip address. So the look up gossip makes in the token metadata to see if the given ip address has any tokens is incorrect.

Second, you say that it is not used to decide whom gossip will talk with, but this contradicts the statement that the node is dropped from gossip. Nodes which are dropped are obviously not talked with. So if we want gossip to be able to continue talking to seeds, we can't drop the seeds from gossip only because gossip can not find tokens which would be associated with the seed ip address.

Reply:
You are confused with two different things. A node gossip can talk with is not required to a node stays in gossip _endpoint_state_map.  In the gossiper::run, gossip can talk to any node, e.g., whatever node specified in the seed list.

You change to `!is_seed(endpoint)` is not needed for the purpose of making gossiper::run be able to talk to nodes in the seed list. 

How to check if a node is part of the ring, using ip or host id is not relevant. When token_metadata is changed to use host id everywhere, gossip can query with host_id as well.

Reply:
Fortunately, @asias I have tests which help me be less confused. The code can declare the node a gossip only member, and then, if it is dead, will proceed with removing it, including from the seed list. Once it's gone, it's gone from everywhere, including the _seeds. So I add a protection from that.

Reply:
This is a wrong protection then. It is ok to drop the node from _endpoint_state_map. If we still want to talk to the nodes specified by the seed option. We can get the seed list from the config directly (which does not change unless user changes it) instead of the _seed in runtime.

Reply:
That would require way more significant changes. Gossiper prepares _seeds set at start and later on is using it for pings. What you suggest either requires we create another copy of this list on stack or as a member of gossiper class (how to properly name this variable to begin with ? _seeds_to_contact_to_fix_kostja_bug?), which would breed confusion, why we have two ways of using the seeds.

Reply:
We already have this to read the configured seed nodes.

```
const std::set<inet_address>& gossiper::get_seeds() const noexcept {
    return _gcfg.seeds;
}       
```

I do not see any confusion.  The logic is very simple, see my comments here:

https://github.com/scylladb/scylladb/pull/12257/files#r1129268139

It is wrong to hack is_gossip_only_member to keep configured seed forever.

Reply:
OK, I don't see why it's wrong, I checked all usages, and changing it the way you suggest will not make any difference. It would be nice if you could provide more rationale to your request. It is unclear why we should have more than one way of working with seeds. Multiplying state creates code that will be even more difficult to maintain down the road, since it would be necessary to keep this state in sync. 
I don't, however, see, any other way to make progress with this than follow your request.

Reply:
I updated the pull request, but I think is_gossip_only_member() will not improve from this. It is undocumented, a seed member is obviously not "gossip only", since it is also a seed, and having two seed lists is not a good idea.